### PR TITLE
Launch/uri in launch spec

### PR DESCRIPTION
**Pull request by KyleGoyette**
Handles the CLI changes required for: https://wandb.atlassian.net/browse/WB-5725

Description
-----------

Adds a WANDB_LAUNCH environment variable. Automatically sets it up in the docker build file. Also sets the WANDB_PROJECT and WANDB_ENTITY in the docker build file.

has wandb_settings skip the project and entity in the init call if in the launch setting.

Testing
-------

How was this PR tested?

### CODE REVIEW

File: wandb/cli/cli.py [hunk 1]
```diff
@@ -936,6 +940,20 @@ def _user_args_to_dict(arguments, argument_type="P"):
     help="Name of the experiment under which to launch the run. If not "
     "specified, 'experiment-id' option will be used to launch run.",
 )
+@click.option(
+    "--entity",
+    "-e",
+    metavar="TEXT",
```
Comment (stephchen) [inline wandb/cli/cli.py:946]: can you add valid metavar https://click.palletsprojects.com/en/5.x/documentation/

File: wandb/cli/cli.py [hunk 2]
```diff
@@ -936,6 +940,20 @@ def _user_args_to_dict(arguments, argument_type="P"):
...
+    default=None,
+    help="Name of the target entity which the new run will be sent to. Defaults to using the entity set by local wandb/settings folder.",
+)
+@click.option(
+    "--project",
+    "-p",
+    metavar="TEXT",
```
Comment (stephchen) [inline wandb/cli/cli.py:953]: same here

File: wandb/sdk/launch/__init__.py [hunk 3]
```diff
@@ -57,7 +55,17 @@ def _run(
     Helper that delegates to the project-running method corresponding to the passed-in backend.
     Returns a ``SubmittedRun`` corresponding to the project run.
     """
-    project = fetch_and_validate_project(uri, experiment_name, api, runner_name, version, entry_point, parameters)
+    project = fetch_and_validate_project(
+        uri, experiment_name, api, runner_name, version, entry_point, parameters
+    )
+
+    if wandb_project is None:
+        wandb_project = api.settings("project")
```
Comment (stephchen) [inline wandb/sdk/launch/__init__.py:63]: what does this default to? is there a project that would exist in the settings

File: wandb/sdk/launch/agent/__init__.py [hunk 4]
```diff
@@ -104,20 +114,28 @@ def run_job(self, job):
...
         run_spec = job["runSpec"]
-        entity = run_spec["entity"]
-        project = run_spec["project"]
-        run_id = run_spec["run_id"]
+        wandb_entity = run_spec["entity"]
+        wandb_project = run_spec["project"]
         resource = run_spec["resource"]
         entry_point = run_spec["overrides"].get("entrypoint")
-        uri = "{}/{}/{}/runs/{}".format(self._base_url, entity, project, run_id)
+        uri = run_spec["uri"]
```
Comment (stephchen) [inline wandb/sdk/launch/agent/__init__.py:123]: lol good catch

File: wandb/sdk_py27/wandb_settings.py [hunk 5]
```diff
@@ -1039,6 +1042,14 @@ def _apply_init(self, args):
                     wandb.termwarn(
                         "Ignored wandb.init() arg %s when running a sweep" % key
                     )
+        if self.launch:
+            for key in ("project", "entity", "id"):
+                val = args.pop(key, None)
+                if val:
+                    wandb.termwarn(
+                        "Ignored wandb.init() arg %s when running running from launch"
```
Comment (stephchen) [inline wandb/sdk_py27/wandb_settings.py:1050]: clearer error message that makes it clear we're only overriding project/entity fields would be good here

**Review by stephchen — COMMENTED** [review]
lgtm just a couple notes and also there's a few debug prints sitting around still. is this tested

File: wandb/sdk/launch/__init__.py (hunk 3 above)
Comment (KyleGoyette) [reply wandb/sdk/launch/__init__.py:63]: So long as there is a `wandb/settings` file the project and entity will be taken from there. Similarly the base_url, which will first be taken from ~/.config/wandb/settings then overwritten by the info in `wandb/settings`.

I'll add an "Uncategorized" option

File: wandb/sdk/launch/agent/__init__.py (hunk 4 above)
Comment (KyleGoyette) [reply wandb/sdk/launch/agent/__init__.py:123]: lmao, some of my initial work may have been that early morning work

**Comment by KyleGoyette** [comment]
Tested with wandb agent wandb launch, changing both the project and using a team entity that I was a member of. I think the right use case of giving a non-team, non-matching with settings user is to error out.

**Pull request merged**